core/src/main/kotlin/McuMgrResult.kt

```diff
@@ -0,0 +1,95 @@
+package com.juul.mcumgr
+
+import com.juul.mcumgr.McuMgrResult.Error
+import com.juul.mcumgr.McuMgrResult.Failure
+import com.juul.mcumgr.McuMgrResult.Success
+import com.juul.mcumgr.message.Response
+
+sealed class McuMgrResult<out T> {
+
+    data class Success<T>(val value: T) : McuMgrResult<T>()
+
+    data class Error<T>(val code: Response.Code) : McuMgrResult<T>()
+
+    data class Failure<T>(val throwable: Throwable) : McuMgrResult<T>()
+
+    val isSuccess: Boolean
+        get() = this is Success
+
+    val isError: Boolean
+        get() = this is Error
+
+    val isFailure: Boolean
+        get() = this is Failure
+
+    fun exceptionOrNull(): Throwable? =
+        when (this) {
+            is Failure -> throwable
+            else -> null
+        }
+}
+
+// On Blocks
+
+inline fun <T> McuMgrResult<T>.onSuccess(action: (response: T) -> Unit): McuMgrResult<T> {
+    when (this) {
+        is Success -> action(value)
+    }
+    return this
+}
+
+inline fun McuMgrResult<*>.onError(action: (code: Response.Code) -> Unit): McuMgrResult<*> {
+    when (this) {
+        is Error -> action(code)
+    }
+    return this
+}
+
+inline fun McuMgrResult<*>.onFailure(action: (throwable: Throwable) -> Unit): McuMgrResult<*> {
+    when (this) {
+        is Failure -> action(throwable)
+    }
+    return this
+}
+
+inline fun McuMgrResult<*>.onErrorOrFailure(action: (throwable: Throwable) -> Unit): McuMgrResult<*> {
+    when (this) {
+        is Error -> action(ErrorCodeException(code))
+        is Failure -> action(throwable)
+    }
+    return this
+}
+
+// Get Or
+
+inline fun <T> McuMgrResult<T>.getOrThrow(): T {
+    return when (this) {
+        is Success -> value
+        is Error -> throw ErrorCodeException(code)
+        is Failure -> throw throwable
+    }
+}
+
+inline fun <T> McuMgrResult<T>.getOrElse(action: (exception: Throwable) -> T): T {
+    return when (this) {
+        is Success -> value
+        is Error -> action(ErrorCodeException(code))
+        is Failure -> action(throwable)
+    }
+}
+
+// Transformations
+
+inline fun <R, T> McuMgrResult<T>.map(transform: (value: T) -> R): McuMgrResult<R> {
+    return when (this) {
+        is Success -> Success(transform(value))
+        is Error -> Error(code)
+        is Failure -> Failure(throwable)
+    }
+}
+
+// Exception
+
+class ErrorCodeException(
+    val code: Response.Code
+) : Exception("Request resulted in error response $code")
```

Comment:
nitpick: Generally better to pick a more specific exception.

Perhaps `IOException` or `IllegalStateException`?